File: backend/kall/models/testimonials.py
from datetime import datetime

from sqlmodel import Field

from kall.clock import utcnow
from kall.models.core import TimestampMixin


class TestimonialRequest(TimestampMixin, table=True):
    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(index=True, foreign_key="user.id")
    recipient_name: str
    recipient_email_encrypted: str
    relationship: str
    request_type: str = "testimonial"
    personal_message: str | None = None
    token_hash: str = Field(index=True, unique=True)
    status: str = "pending"
    expires_at: datetime | None = None
    sent_at: datetime | None = None
    completed_at: datetime | None = None
    revoked_at: datetime | None = None


class Testimonial(TimestampMixin, table=True):
    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(index=True, foreign_key="user.id")
    request_id: int | None = Field(default=None, foreign_key="testimonialrequest.id")
    author_name: str
    author_title: str | None = None
    author_company: str | None = None
    relationship: str
    body: str
    status: str = "pending_review"
    verified_via_request: bool = False
    include_on_profile: bool = False
    include_in_applications: bool = False
    permission_granted: bool = False
    approved_at: datetime | None = None
    withdrawn_at: datetime | None = None
    withdrawal_reason: str | None = None


class ApplicationTestimonial(TimestampMixin, table=True):
    id: int | None = Field(default=None, primary_key=True)
    application_id: int = Field(index=True, foreign_key="application.id")
    testimonial_id: int = Field(index=True, foreign_key="testimonial.id")
    user_id: int = Field(index=True, foreign_key="user.id")
    included_at: datetime = Field(default_factory=utcnow)

File: apps/mobile/src/api/testimonials.ts
import { apiRequest } from "./client";

export type Testimonial = {
  id: number;
  request_id: number | null;
  author_name: string;
  author_title: string | null;
  author_company: string | null;
  relationship: string;
  body: string;
  status: "pending_review" | "approved" | "rejected" | "withdrawn" | string;
  verified_via_request: boolean;
  include_on_profile: boolean;
  include_in_applications: boolean;
  permission_granted: boolean;
  approved_at: string | null;
  withdrawn_at: string | null;
  created_at: string;
};

export type TestimonialRequest = {
  id: number;
  recipient_name: string;
  relationship: string;
  request_type: string;
  status: string;
  expires_at: string | null;
};

export const fetchTestimonials = () => apiRequest<Testimonial[]>("/testimonials");

/** The invitation token comes back exactly once; nothing can re-fetch it. */
export const createTestimonialRequest = (body: {
  recipient_name: string;
  recipient_email: string;
  relationship: string;
  request_type: string;
  personal_message: string | null;
}) => apiRequest<{ request: TestimonialRequest; invitation_token: string }>("/testimonials/requests", { method: "POST", body });

export const revokeTestimonialRequest = (requestId: number) =>
  apiRequest<{ status: string }>(`/testimonials/requests/${requestId}`, { method: "DELETE" });

export const moderateTestimonial = (
  id: number,
  body: { status: "approved" | "rejected" | "pending_review"; include_on_profile: boolean; include_in_applications: boolean },
) => apiRequest<Testimonial>(`/testimonials/${id}`, { method: "PUT", body });
